rbd: add validation for thick restore/clone

Madhu-1 opened the pull request:
added validation to allow only Restore of Thick PVC snapshot to a thick clone and creation of a thick clone from thick PVC.
    
Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: make isThickProvisioned method of rbdImage

isThickProvisioned can be used for both snapshot
and clone validation if isThickProvisioned is method
of common rbdImage structure.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: add validation for thick restore/clone

added validation to allow only Restore of Thick PVC
snapshot to a thick clone and creation of thick clone
from thick PVC.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

nixpanic commented:
In theory we can also call [`rbd_sparsify()`](https://pkg.go.dev/github.com/ceph/go-ceph/rbd#Image.Sparsify) on the new image if it should not be thick-provisioned.

Maybe there should be an issue to track what we actually want to do, and why?

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -1570,3 +1570,19 @@ func (ri *rbdImage) isCompatibleEncryption(dst *rbdImage) error {
+	}
+	switch {
+	case thick && !dst.ThickProvision:
+		return fmt.Errorf("cannot create thin volume from thick volume %q", ri)
```
The error message mentions creating, but that may not always be correct. use "does not match" or similar instead?

Madhu-1 replied:
keeping error message same as https://github.com/ceph/ceph-csi/pull/2175 any suggestion on the error message?

nixpanic replied:
I do not care too much about it :) for now it is only used while creating volumes, so that's fine

humblec replied:
I think current message is fine: we can also change to  something like this:
`Source ( thin/thick) : Destination ( thin/thick) - Unsupported conversion`:

and fill `thin` and `thick` strings accordingly in these cases

Madhu-1 replied:
Thanks, @humblec as this is a clear error message and fine let's keep it as it is for now. If required we can change it later.

Madhu-1 commented:
> In theory we can also call [`rbd_sparsify()`](https://pkg.go.dev/github.com/ceph/go-ceph/rbd#Image.Sparsify) on the new image if it should not be thick-provisioned.
> 
> Maybe there should be an issue to track what we actually want to do, and why?

To keep the feature simple for now let's allow only thick-thick will open one for this one and will open another issue to track thick-thin and thin-thick feature

humblec commented:
@Madhu-1 please let me know your choice to accept or defer https://github.com/ceph/ceph-csi/pull/2174#discussion_r653297992, accordingly we can move :)

Madhu-1 commented:
> @Madhu-1 please let me know your choice to accept or defer [#2174 (comment)](https://github.com/ceph/ceph-csi/pull/2174#discussion_r653297992), accordingly we can move :)

Already replied at https://github.com/ceph/ceph-csi/pull/2174#discussion_r653298945 :)

Madhu-1 committed:
rbd: use RbdSnapName to check the image details

RbdSnapName holds the actual RBD image name which
got created during the CreateSnapshot operation.
RbdImageName holds the name of the parent from
which the snapshot is created. and the parent
is independent of snapshot and it can be deleted
any time for the same reason using the RbdSnapName
to check the rbd image details.

generate a temporary volume from the snapshot which
replaces the rbdImageName with RbdSnapName and use
it to check the image metadata.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 commented:
@nixpanic @humblec added one more commit to fix some cases where parent PVC of the snapshot is created before Restore operation. PTAL

Madhu-1 commented:
/retest ci/centos/upgrade-tests-cephfs

Madhu-1 commented:
> /retest ci/centos/upgrade-tests-cephfs

` Error from server: error when creating "STDIN": etcdserver: request timed out`

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -218,15 +218,23 @@ func checkValidCreateVolumeRequest(rbdVol, parentVol *rbdVolume, rbdSnap *rbdSna
 
 	switch {
 	case rbdSnap != nil:
+
```
drop this new line?

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -218,15 +218,23 @@ func checkValidCreateVolumeRequest(rbdVol, parentVol *rbdVolume, rbdSnap *rbdSna
 		if err != nil {
 			return status.Errorf(codes.InvalidArgument, "cannot restore from snapshot %s: %s", rbdSnap, err.Error())
 		}
+		// During CreateSnapshot the rbd image will be created with the
```
wow, this is really ugly :cry:

Unfortunatelty I think I understand why this is needed. However, I prefer to have a `isCompatibleThickProvision()` function for the `rbdSnapshot` type instead. That can use the `RbdSnapName` for checking the thick-provisioning state.

(temporary) changing attributes of objects should not be done, it shows lack of proper usage/design of the type in many cases. The `rbdSnapshot` type should probably get corrected, and once that is done, the duplicate `isCompatibleThickProvision()` function can be removed/reconciled.

Madhu-1 replied:
Pushed a new commit with a better approach instead of changing attributes on the snapshot we can generate a new temp volume and use it for checking the metadata and we can destroy it once work is done.

humblec replied:
@Madhu-1 I am not favoring  the latest approach of  an intermediate volume. The reason being an image struct creation in the backend just to hide our code issue :) ..

Madhu-1 replied:
@humblec please let me know your suggestion i can do the same.

Madhu-1 replied:
we are not creating an image in the backend it's just a generation of the new temp volume structure.

humblec replied:
yes, its just a generation of the new struct, but previously we were flipping a field in the struct, now we are completely creating a new structure.. I am wondering how this favors the other :).. anyway lets get one more opinion on this and decide.

Madhu-1 replied:
This approach covers the original concern without modifying the original rbdSnapshot struct and reverts it back. the current temp volume generation is already used in other places, not something new.

humblec replied:
yep, the entire code path need to be revisited in some time. However considering the priority lets survive with this model. Thanks

Madhu-1 committed:
rbd: set thick provision metadata on clone volume

the parent volume(CreateVolume) and the clone volume
(CreateSnapshot) are both indepedent and parent volume
can be deleted anytime. To check the thick provision
during Snapshot restore(CreateVolume from snapshot)
we need the thick provision metadata so for the same
reason setting the thick provision metadata on the
clone image we are creating at the CreateSnapshot time.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

humblec commented on internal/rbd/controllerserver.go:
```diff
@@ -860,6 +870,30 @@ func cloneFromSnapshot(ctx context.Context, rbdVol *rbdVolume, rbdSnap *rbdSnaps
+	// The clone image created during CreateSnapshot has to be marked as thick.
+	// As snapshot and volume both are independent we cannot depend on the
+	// parent volume of the clone to check thick provision during CreateVolume
+	// From snapshot operation because the parent volume can be deleted anytime
```
s/From/from

humblec commented on internal/rbd/controllerserver.go:
```diff
@@ -951,6 +985,24 @@ func (cs *ControllerServer) doSnapshotClone(ctx context.Context, parentVol *rbdV
+	// The clone image created during CreateSnapshot has to be marked as thick.
+	// As snapshot and volume both are independent we cannot depend on the
+	// parent volume of the clone to check thick provision during CreateVolume
+	// From snapshot operation because the parent volume can be deleted anytime
```
s/From/from

humblec commented on internal/rbd/rbd_util.go:
```diff
@@ -1570,3 +1570,47 @@ func (ri *rbdImage) isCompatibleEncryption(dst *rbdImage) error {
+func (rs *rbdSnapshot) isCompatibleThickProvision(dst *rbdVolume) error {
+	// During CreateSnapshot the rbd image will be created with the
+	// snapshot name. Replacing RbdImageName with RbdSnapName so that we
+	// can check if the image is t	hick provisioned
```
typo in thick above.

Madhu-1 commented:
/test ci/centos/upgrade-tests-rbd

Madhu-1 commented:
> /test ci/centos/upgrade-tests-rbd

`Error from server: error when creating "STDIN": etcdserver: request timed out`

Madhu-1 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.19

Madhu-1 commented:
> /retest ci/centos/mini-e2e-helm/k8s-1.19

`failed to create PVC with error etcdserver: request timed out`
